onnxruntime/python/tools/transformers/models/bart/utils/export_summarization_enc_dec_past.py

```diff
@@ -0,0 +1,241 @@
+import copy
+import os
+import time
+
+import torch
+from transformers import BartForConditionalGeneration, file_utils
+from utils import export_helper
+
+from onnxruntime import InferenceSession, SessionOptions
+
+
+def decoder_config_update(config):
+    """
+    Add parameters into decoder config to help expoerting. These parameter
+    are later consumed with control flow and assertion
+
+    Args:
+        config: BART config
+    """
+
+    new_attributes = {
+        "is_decoder_exported": False,
+        "is_decoder_with_past_exported": False,
+        "during_export": False,
+        "expected_args": {"use_cache": True, "return_dict": True},
+        "expected_inputs": ["decoder_input_ids", "attention_mask", "encoder_outputs"],
+    }
+    config.update(new_attributes)
+    return config
+
+
+class DecoderWrapper(torch.nn.Module):
+    def __init__(self, m):
+        super().__init__()
+        self.m = m
+
+    def forward(self, decoder_input_ids, attention_mask, encoder_hidden_states, *past):
+        encoder_outputs = file_utils.ModelOutput()
+        encoder_outputs["last_hidden_state"] = encoder_hidden_states
+        encoder_outputs["hidden_states"] = None
+        encoder_outputs["attentions"] = None
+        if len(past) == 0:
+            past_key_values = None
+        else:
+            past_key_values = export_helper.PastKeyValuesHelper.back_group_by_layer(past)
+        decoder_out = self.m(
+            None,
+            encoder_outputs=encoder_outputs,
+            decoder_input_ids=decoder_input_ids,
+            attention_mask=attention_mask,
+            past_key_values=past_key_values,
+            use_cache=True,
+            return_dict=True,
+        )
+        present_self, _ = export_helper.PastKeyValuesHelper.group_by_self_and_cross(decoder_out.past_key_values)
+        return decoder_out.logits, present_self
+
+
+def _decoder_forward_wrapper(model, decoder_config, args):
+    """
+    Exporting the decoder of BART by inserting torch.onnx.export and wrapper
+    torch.nn.Module into `forward`.
+
+    Args:
+        model: BartForConditionalGeneration
+        args: User input
+        decoder_config: BART config after `decoder_config_update`
+    """
+    model._forward = model.forward
+
+    def forward(self, input_ids, attention_mask, **kwargs):
+        # attention_mask is required to show as forward parameters based on `generation_utils.py`
+        # reference: https://github.com/huggingface/transformers/blob/6faf283288ce3390281ad8c1d37ccb13f2d03990/src/transformers/generation_utils.py#L1200-L1203
+        kwargs["attention_mask"] = attention_mask
+
+        if decoder_config.is_decoder_exported and decoder_config.is_decoder_with_past_exported:
+            outputs = self._forward(input_ids, **kwargs)
+            return outputs
+
+        inputs = []
+        for k, v in decoder_config.expected_args.items():
+            assert k in kwargs, f"Expecting {k} in kwargs."
```

Comment:
Are kwargs provided by users? If so raise a ValueError instead of using assert for user inputs

Use assert only for internal invariants

Reply:
It's actually inner model parameters. Assert makes more sense, and doesn't break exporter.

Reply:
> It's actually inner model parameters (you get those key value from previous iteration calculation). Assert makes more sense, and doesn't break exporter.